### PR TITLE
Add eval examples

### PR DESCRIPTION
**Pull request by Hakuyume**

### CODE REVIEW

**Comment by yuyu2172** [comment]
Why did you put the function under `utils/testing/`?
I do not think this is only for testing.

Do you think `evaluations/eval_detction_mode.py` is a bad idea?

Also, `eval.py` can be changed to take both Faster R-CNN and SSD.

**Comment by Hakuyume** [comment]
> Why did you put the function under utils/testing/?
I do not think this is only for testing.

As you pointed out , this not only for testing. I agree this function should be moved.

> Do you think evaluations/eval_detction_mode.py is a bad idea?

I think `evaluations.eval_detection_link_voc` is a good idea. However, `apply_detection_link` is not only for evaluation (for example, it can be used for visualization and testing images without ground truth). So I don't want to put `apply_detection_link` under `evaluations`.

**Comment by Hakuyume** [comment]
> Also, eval.py can be changed to take both Faster R-CNN and SSD.

Yes, I think this is a good idea. How about putting `eval.py` and `visualize_detection_models.py` into `examples/detection`?

**Comment by yuyu2172** [comment]
> I think evaluations.eval_detection_link_voc is a good idea. However, apply_detection_link is not only for evaluation (for example, it can be used for visualization and testing images without ground truth). So I don't want to put apply_detection_link under evaluations.

How about `chainercv/iterators/apply_detection_link`.
I think there will be more functions that take iterators as inputs.

**Comment by yuyu2172** [comment]
> Yes, I think this is a good idea. How about putting eval.py and visualize_detection_models.py into examples/detection?

Yeah. I think that is good.

**Comment by Hakuyume** [comment]
> How about chainercv/iterators/apply_detection_link.
I think there will be more functions that take iterators as inputs.
Also, this directory will imply that the detecion_link is going to be applied on iterators.

In Chainer, `chainer/iterators` is a collection of `Iterator`s. How about `chainercv/utils/iterator`?

**Comment by yuyu2172** [comment]
> In Chainer, chainer/iterators is a collection of Iterators. How about chainercv/utils/iterator?

That sounds good!

File: chainercv/utils/iterator/apply_detection_link.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    into three lists,
+    :obj:`pred_bboxes`, :obj:`pred_labels` and :obj:`pred_scores`.
+    This function also stacks the values returned by the iterator.
+    These values can be used for evaluation.
+
+    Args:
+        target (chainer.Link): An detection link. This link must have
+            :meth:`predict` method which take a list of images and returns
+            :obj:`bboxes`, :obj:`labels` and :obj:`scores`.
+        iterator (chainer.Iterator): An iterator. Each samples must have
```
Comment (yuyu2172) [inline chainercv/utils/iterator/apply_detection_link.py]: s/samples/sample

By the way, I think "should" sounds friendlier than "must".

File: chainercv/utils/iterator/apply_detection_link.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+            an image for its first element. This image is passed to
```
Comment (yuyu2172) [inline chainercv/utils/iterator/apply_detection_link.py]: an image as its first elemnt

File: chainercv/utils/iterator/apply_detection_link.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+            :obj:`target`. The rests are stacked into :obj:`gt_values`.
+        hook: An callable which called after each iteration.
```
Comment (yuyu2172) [inline chainercv/utils/iterator/apply_detection_link.py]: which is

File: chainercv/utils/iterator/apply_detection_link.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            :obj:`pred_bboxes`, :obj:`pred_labels`, :obj:`pred_scores` and
+            :obj:`gt_values` are passed. Note that these values do not contain
```
Comment (yuyu2172) [inline chainercv/utils/iterator/apply_detection_link.py]: are passed as arguments

File: chainercv/utils/iterator/apply_detection_link.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+            the data of previous iterations.
```
Comment (yuyu2172) [inline chainercv/utils/iterator/apply_detection_link.py]: data from the previous iterations

File: chainercv/utils/iterator/apply_detection_link.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+    returns:
```
Comment (yuyu2172) [inline chainercv/utils/iterator/apply_detection_link.py]: s/returns/Returns/

File: chainercv/utils/iterator/apply_detection_link.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        tuple of list:
+        This function returns four lists, :obj:`pred_bboxes`,
```
Comment (yuyu2172) [inline chainercv/utils/iterator/apply_detection_link.py]: s/four lists,/four lists:

File: examples/detection/README.md [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+- SSD512
+
+For the details, please check the documents and examples of each model.
+
+## Performance
+
+### PASCAL VOC2007 Test
+
+| Model | FPS | mAP (PASCAL VOC2007 metric) |
+|:-:|:-:|:-:|
```
Comment (yuyu2172) [inline examples/detection/README.md]: perhaps, it is better to add a column on training data.

**Comment by yuyu2172** [comment]
How about adding a doc rst

**Comment by yuyu2172** [comment]
Some grammar comments are not reflected yet.

File: chainercv/utils/iterator/apply_detection_link.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        iterator (chainer.Iterator): An iterator. Each sample should have
+            an image as its first element. This image is passed to
+            :obj:`target`. The rests are stacked into :obj:`gt_values`.
+        hook: An callable which is called after each iteration.
+            :obj:`pred_bboxes`, :obj:`pred_labels`, :obj:`pred_scores` and
+            :obj:`gt_values` are passed. Note that these values do not contain
+            the data of previous iterations.
+
+    Returns:
+        tuple of list:
```
Comment (yuyu2172) [inline chainercv/utils/iterator/apply_detection_link.py]: tuple of lists

**Comment by yuyu2172** [comment]
LGTM

**Pull request merged**